hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java

```diff
@@ -40,96 +44,345 @@
+
+    if (numDatanodesBalanced + numDatanodesToBalance > maxDatanodesToBalance) {
+      LOG.info("Approaching Max Datanodes To Balance limit in Container " +
+          "Balancer. Stopping Balancer.");
+      stop();
+      return false;
+    } else {
+      unBalancedNodes.addAll(overUtilizedNodes);
+      unBalancedNodes.addAll(underUtilizedNodes);
+
+      if (unBalancedNodes.isEmpty()) {
+        LOG.info("Did not find any unbalanced Datanodes.");
+        stop();
+        return false;
+      } else {
+        LOG.info("Container Balancer has identified Datanodes that need to be" +
+            " balanced.");
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Performs binary search to determine if the specified listToSearch
+   * contains the specified node.
+   *
+   * @param listToSearch List of DatanodeUsageInfo to be searched.
+   * @param node DatanodeUsageInfo to be searched for.
+   * @return true if the specified node is present in listToSearch, otherwise
+   * false.
+   */
+  private boolean containsNode(
+      List<DatanodeUsageInfo> listToSearch, DatanodeUsageInfo node) {
+    int index = 0;
+    Comparator<DatanodeUsageInfo> comparator =
+        DatanodeUsageInfo.getMostUsedByRemainingRatio();
+
+    if (comparator.compare(listToSearch.get(0),
+        listToSearch.get(listToSearch.size() - 1)) < 0) {
+      index =
+          Collections.binarySearch(listToSearch, node, comparator.reversed());
+    } else {
+      index = Collections.binarySearch(listToSearch, node, comparator);
+    }
+    return index >= 0 && listToSearch.get(index).equals(node);
+  }
+
+  /**
+   * Calculates the number of used bytes given capacity and utilization ratio.
+   *
+   * @param nodeCapacity capacity of the node.
+   * @param utilizationRatio used space by capacity ratio of the node.
+   * @return number of bytes
+   */
+  private double ratioToBytes(Long nodeCapacity, double utilizationRatio) {
+    return nodeCapacity * utilizationRatio;
+  }
+
+  /**
+   * Calculates the average datanode utilization for the specified nodes.
+   * Utilization is used space divided by capacity.
+   *
+   * @param nodes List of DatanodeUsageInfo to find the average utilization for
+   * @return Average utilization value
+   * @throws ArithmeticException Division by zero
+   */
+  private double calculateAvgUtilization(List<DatanodeUsageInfo> nodes)
+      throws ArithmeticException {
     SCMNodeStat aggregatedStats = new SCMNodeStat(
         0, 0, 0);
     for (DatanodeUsageInfo node : nodes) {
       aggregatedStats.add(node.getScmNodeStat());
     }
-    return aggregatedStats.getScmUsed().get().doubleValue() /
-        aggregatedStats.getCapacity().get().doubleValue();
+    clusterCapacity = aggregatedStats.getCapacity().get();
+    clusterUsed = aggregatedStats.getScmUsed().get();
+    clusterRemaining = aggregatedStats.getRemaining().get();
+
+    try {
+      return clusterUsed / (double) clusterCapacity;
+    } catch (ArithmeticException e) {
```

Comment:
NIT. better not handle `ArithmeticException`, instead check `nodes.size() != 0` at the entrance of function.

Reply:
Updated PR. Please take a look.